Validate storage account type against attribute bindings

mamaso opened the pull request:
related to #810 and [script/#773](https://github.com/Azure/azure-webjobs-sdk-script/issues/773)

Looking for initial feedback before I add tests & polish.

Added `StorageAccountType` to storage accounts, which is set to `premium` or `blobonly` in `validateCredentialsAsync`.  ~~In each binding provider I test if the storage account type is compatible with that attribute.~~

When you request a StorageAccount you supply the supported storage account types.  If the account is not of the supported type, we throw an error indicating the invalid type & what types are supported.

 Added `SupportedTypes` to StorageAccountAttribute with default `All`, so that from functions we can disabled Premium storage for BlobTriggerAttributes.

``` c#
// to throw if 'somename' is a Premium storage account, you can write this:
StorageAccount('somename', ~StorageAccountTypes.Premium)
```

paulbatum commented on src/Microsoft.Azure.WebJobs.Host/Blobs/Triggers/BlobTriggerAttributeBindingProvider.cs:
```diff
@@ -160,6 +160,11 @@ private static IBlobArgumentBindingProvider CreateProvider(IEnumerable<Type> clo
             IStorageAccount hostAccount = await _accountProvider.GetStorageAccountAsync(context.CancellationToken);
             IStorageAccount dataAccount = await _accountProvider.GetStorageAccountAsync(context.Parameter, context.CancellationToken, _nameResolver);
 
+            if (dataAccount.Type == StorageAccountType.Premium)
```
It makes more sense for this logic to be applied at the ScriptHost level. At the webjobs level there is no problem with the trigger.

paulbatum commented on src/Microsoft.Azure.WebJobs.Host/Executors/DefaultStorageCredentialsValidator.cs:
```diff
@@ -56,40 +54,42 @@ internal class DefaultStorageCredentialsValidator : IStorageCredentialsValidator
-                IStorageQueue queue = queueClient.GetQueueReference("name");
-                try
+                var storageException = e as StorageException;
+                if (storageException != null && storageException.RequestInformation.HttpStatusCode == 400 &&
```
This needs some explanatory comments :)

paulbatum commented on src/Microsoft.Azure.WebJobs.Host/Executors/DefaultStorageCredentialsValidator.cs:
```diff
@@ -56,40 +54,42 @@ internal class DefaultStorageCredentialsValidator : IStorageCredentialsValidator
+            catch (StorageException exception)
+            {
+                WebException webException = exception.GetBaseException() as WebException;
+                if (webException != null && webException.Status == WebExceptionStatus.NameResolutionFailure)
```
Same as above, needs some //WARNING: HACK INCOMING comments.

mathewc commented on src/Microsoft.Azure.WebJobs/StorageAccountAttribute.cs:
```diff
@@ -27,14 +27,22 @@ public sealed class StorageAccountAttribute : Attribute
+        /// <summary>
+        /// Gets the disabled storage account types.
+        /// </summary>
+        public StorageAccountTypes Supported { get; private set; }
```
Let's not complicate the public UI for people with this - they shouldn't have to tell us what the account type is. I don't want our inability to determine a storage account type to leak through, since that will be temporary - we need to get a way to do this from storage (which we should/will be able to do eventually). We should push instead for that.

mamaso replied:
The motivation is to allow functions to disable blobtriggers on premium.  I agree with keeping the public api simple (this probably doesn't have much value to an average joe user).  

Is it preferable to have a way of detecting whether the web jobs sdk is running in a functions environment?  @fabiocav argued against this, hence the attribute.

mamaso replied:
@mathewc Any preference on how to proceed?  Look for existence of FUNCTIONS_EXTENSION_VERSION and disable BlobTrigger on premium accounts?

I think it's either that or surface a method (like the storage account attribute) so that script code can control whether these are enabled, unless if you see another option.  Is there a global configuration for the core sdk?

mamaso commented:
@mathewc @fabiocav implemented the stuff discussed yesterday.

- Removed all public api stuff: StorageAccountType enum, StorageAccountAttribute 'SupportedTypes' property
- Disabled BlobTrigger for premium storage accounts at the webjobs level
- Added / fixed tests

Behavior:  Throws an exception like `Storage account 'name' is of unsupported type 'Premium'. Storage account must be of type(s) GeneralPurpose, BlobOnly` upon job host start

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Executors/DefaultStorageCredentialsValidator.cs:
```diff
@@ -56,40 +55,52 @@ internal class DefaultStorageCredentialsValidator : IStorageCredentialsValidator
                 {
-                    await queue.ExistsAsync(cancellationToken);
-                    _primaryCredentials = account.Credentials;
+                    // WARNING: HACK INCOMING
```
Can you change this warning to a TODO pointing to a new bug number, tracking getting the storage team to give us a better way of identifying this? I believe you mentioned there was an issue in their repo related to this.

mamaso replied:
sure thing

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Executors/DefaultStorageCredentialsValidator.cs:
```diff
@@ -56,40 +55,52 @@ internal class DefaultStorageCredentialsValidator : IStorageCredentialsValidator
+                    // WARNING: HACK INCOMING
+                    // Blob-only storage accounts do not support services other than Blob.  
+                    // If we see a name resolution failure on the queue endpoint classify as a blob-only account
+                    account.Type = StorageAccountTypes.BlobOnly;
```
Should this method default the Type to GeneralPurpose at the top of the method?

mamaso replied:
Perhaps for clarity.  I just had every storage account constructed as GeneralPurpose by default, but it probably is useful to show that here

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Executors/DefaultStorageCredentialsValidator.cs:
```diff
@@ -56,40 +55,52 @@ internal class DefaultStorageCredentialsValidator : IStorageCredentialsValidator
-                        throw new InvalidOperationException(message);
-                    }
-                    throw;
+                    // WARNING: HACK INCOMING
```
Why do you need this HACK comment here? I'd just remove it. We're not planning on having a better way here, right?

mamaso replied:
the better way is storage api to determine account type, yeah hack comment is unnecessary

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Executors/StorageAccountProviderExtensions.cs:
```diff
@@ -19,7 +19,7 @@ public static Task<IStorageAccount> GetDashboardAccountAsync(this IStorageAccoun
             }
 
-            return provider.GetAccountAsync(ConnectionStringNames.Dashboard, cancellationToken);
+            return provider.GetAccountAsync(ConnectionStringNames.Dashboard, StorageAccountTypes.GeneralPurpose, cancellationToken);
```
I think you should have a default parameter for the account type, so for the mainline case you don't have to modify the code everywhere. We should only have to pass in this parameter in cases where you need to restrict.

mamaso replied:
I agree, in the past I've gotten flak for not having cancellation token be the last parameter so I did it this way this time.  I'm all for cleaner code so I'll use a default.

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Executors/StorageAccountProviderExtensions.cs:
```diff
@@ -54,7 +50,7 @@ public static Task<IStorageAccount> GetStorageAccountAsync(this IStorageAccountP
             }
 
-            return provider.GetAccountAsync(connectionStringName, cancellationToken);
+            return provider.GetAccountAsync(connectionStringName, bindingSupported, cancellationToken);
```
I'd rename this parameter to accountTypeSupported?

mamaso replied:
good point, left over from when I was doing attributeSupported vs bindingSupported stuff

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Executors/DefaultStorageAccountProvider.cs:
```diff
@@ -169,14 +168,21 @@ private IStorageAccount StorageAccount
-                await _storageCredentialsValidator.ValidateCredentialsAsync(account, isPrimary, cancellationToken);
+                await _storageCredentialsValidator.ValidateCredentialsAsync(account, cancellationToken);
+
+                if (!supportedTypes.HasFlag(account.Type))
```
How about we move this out into its own static helper method, since the logic is separate from validating/classifying the account. E.g. perhaps an extension method to IStorageAccount.

That way, you don't have to modify all of the existing function signatures to pass in the type. I think there are only 2 places where you're actually doing an account restriction check anyways - so you can call to the new method in those places only.

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Storage/StorageAccountType.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    /// <summary>Storage account types.</summary>
+    [Flags]
+    internal enum StorageAccountTypes
```
I don't think this enum needs to be a flags. Why can't it be named StorageAccountType, with GeneralPurpose, BlobOnly, Premium? For the checks you need to do, I don't think you need flags.

mamaso replied:
Arose from the storage account attribute stuff.  I'd still say it's useful because you need to pass a list of enums around to do the below matrix, unless if we have extension methods for 'onlyAllow' and 'doNotAllow' (which would not scale if they added another account type with weird restrictions that didnt' fall in these buckets).

- Storage/Dashboard accounts / table bindings / queue bindings == GeneralPurpose
- Blob Triggers / SingletonManager == General Purpose || BlobOnly
- Blob input / output == all types

mathewc commented:
Much cleaner now. :ship:
